docs/source/schema-directives.md

```diff
@@ -1,179 +1,581 @@
+
+```js
+const typeDefs = `
+directive @auth(
+  requires: Role = ADMIN,
+) on OBJECT | FIELD_DEFINITION
+
+enum Role {
+  ADMIN
+  REVIEWER
+  USER
+  UNKNOWN
+}
+
+type User @auth(requires: USER) {
+  name: String
+  banned: Boolean @auth(requires: ADMIN)
+  canPost: Boolean @auth(requires: REVIEWER)
+}`;
+
+// Symbols can be a good way to store semi-hidden data on schema objects.
+const authRoleSymbol = Symbol.for("@auth role");
+const authWrapSymbol = Symbol.for("@auth wrapped");
+
+class AuthDirective extends SchemaDirectiveVisitor {
+  visitObject(type) {
+    this.ensureFieldsWrapped(type);
+    type[authRoleSymbol] = this.args.requires;
+  }
+
+  visitFieldDefinition(field, details) {
+    this.ensureFieldsWrapped(details.objectType);
+    field[authRoleSymbol] = this.args.requires;
+  }
+
+  ensureFieldsWrapped(type) {
+    // Mark the GraphQLObjectType object to avoid re-wrapping its fields:
+    if (type[authWrapSymbol]) {
+      return;
+    }
+
+    const fields = type.getFields();
+    Object.keys(fields).forEach(fieldName => {
+      const field = fields[fieldName];
+      const { resolve = defaultFieldResolver } = field;
+      field.resolve = async function (...args) {
+        // Get the required role from the field first, falling back to the
+        // parent GraphQLObjectType if no role is required by the field:
+        const requiredRole = field[authRoleSymbol] || type[authRoleSymbol];
+        if (! requiredRole) {
+          return resolve.apply(this, args);
+        }
+        const context = args[2];
+        const user = await getUser(context.headers.authToken);
+        if (! user.hasRole(requiredRole)) {
+          throw new Error("not authorized");
+        }
+        return resolve.apply(this, args);
+      };
+    });
+
+    type[authWrapSymbol] = true;
+  }
 };
 
-// Required: Export the GraphQL.js schema object as "schema"
-export const schema = makeExecutableSchema({
+const schema = makeExecutableSchema({
   typeDefs,
-  resolvers,
-  directiveResolvers,
+  directives: {
+    auth: AuthDirective,
+    authorized: AuthDirective,
+    authenticated: AuthDirective
+  }
 });
 ```
 
-The result with query `{foo}` will be:
-```json
-{
-  "data": {
-    "foo": "FOO@GMAIL.COM"
+### Enforcing value restrictions
+
+Suppose you want to enforce a maximum length for a string-valued field:
+
+```js
+const typeDefs = `
+directive @length(max: Int) on FIELD_DEFINITION | INPUT_FIELD_DEFINITION
+
+type Query {
+  books: [Book]
+}
+
+type Book {
+  title: String @length(max: 50)
+}
+
+type Mutation {
+  createBook(book: BookInput): Book
+}
+
+input BookInput {
+  title: String! @length(max: 50)
+}`;
+
+class LengthDirective extends SchemaDirectiveVisitor {
+  visitInputFieldDefinition(field) {
+    this.wrapType(field);
+  }
+
+  visitFieldDefinition(field) {
+    this.wrapType(field);
+  }
+
+  wrapType(field) {
```

Comment:
It took me a minute to realize that this actually replaces the field type with a custom scalar - maybe we should add a comment to show that's what is happening.

Also, this might have undesirable side effects, for example:

```graphql
mutation ($bookTitle: String!) {
  createBook(book: { title: bookTitle }) { title }
}
```

After this transformation it will no longer be a valid mutation because the `String` type won't match. So it might be better to actually wrap the resolver and do the check at runtime.

Reply:
How would that work with input fields?